src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -160,7 +168,41 @@ static IContainSagaData DeepClone(IContainSagaData source)
 
             public IContainSagaData GetSagaCopy()
             {
-                return DeepClone(data);
+                var canBeClonedShallowly = canBeClonedShallowlyCache.GetOrAdd(data.GetType(), CanBeMemberwiselyCloned);
+                return canBeClonedShallowly ? memberwiseClone(data) : DeepClone(data);
```

Comment:
If you rename `DeepClone` to `DeepCopy`, with the other suggested changes this becomes:

`return canBeShallowCopied ? shallowCopy(data) : DeepCopy(data);`

which I think works better.